contracts/beyondFramework/dca/base/DCABaseUpgradeable.sol

```diff
@@ -0,0 +1,1168 @@
+        setSwapPath(
+            args.depositToBluechipSwapPath,
+            args.bluechipToDepositSwapPath
+        );
+    }
+
+    modifier onlyDcaInvestor() {
+        require(_msgSender() == dcaInvestor, "Unauthorized");
+        _;
+    }
+
+    modifier onlyPortfolio() {
+        bool authorized;
+        for (uint256 i = 0; i < portfolios.length; i++) {
+            if (portfolios[i] == _msgSender()) {
+                authorized = true;
+                break;
+            }
+        }
+
+        require(authorized, "Unauthorized");
+        _;
+    }
+
+    modifier nonEmergencyExited() {
+        require(
+            bluechipInvestmentState != BluechipInvestmentState.EmergencyExited,
+            "Strategy is emergency exited"
+        );
+
+        _;
+    }
+
+    modifier emergencyWithdrawOnEmergencyExitedStatus(address sender) {
+        // if emergency exited then user should receive everything in deposit asset
+        if (isEmergencyExited()) {
+            _emergencyWithdrawUserDeposit(sender);
+            return;
+        }
+
+        _;
+    }
+
+    receive() external payable {}
+
+    // ----- Base Class Methods -----
+    function deposit(uint256 amount, uint8 amountSplit)
+        public
+        virtual
+        nonReentrant
+        whenNotPaused
+        nonEmergencyExited
+    {
+        _deposit(_msgSender(), amount, amountSplit);
+    }
+
+    function depositFor(
+        address sender,
+        uint256 amount,
+        uint8 amountSplit
+    )
+        public
+        virtual
+        onlyPortfolio
+        nonReentrant
+        whenNotPaused
+        nonEmergencyExited
+    {
+        _deposit(sender, amount, amountSplit);
+    }
+
+    function _deposit(
+        address sender,
+        uint256 amount,
+        uint8 amountSplit
+    ) private {
+        // assert valid amount sent
+        if (amount < minDepositAmount) {
+            revert TooSmallDeposit();
+        }
+
+        // transfer deposit token from portfolio
+        depositTokenInfo.token.safeTransferFrom(
+            _msgSender(),
+            address(this),
+            amount
+        );
+
+        // compute actual deposit and transfer fee to receiver
+        amount = _takeFee(amount);
+
+        DcaDepositor storage depositor = depositors[sender];
+
+        // assert positions limit is not reached
+        if (depositor.positions.length == positionsLimit) {
+            revert PositionsLimitReached();
+        }
+
+        // add splitted amounts to the queue
+        globalInvestQueue.splitUserInvestmentAmount(amount, amountSplit);
+
+        // if not started position with the same split exists - increase deposit amount
+        for (uint256 i = 0; i < depositor.positions.length; i++) {
+            // calculate amount of passed investment epochs
+            uint256 passedInvestPeriods = (lastInvestmentTimestamp -
+                depositor.positions[i].investedAt) / investmentPeriod;
+
+            if (
```

Comment:
if someone invested for 10 periods, and 2 periods have passed, then if he reinvest now with 8 period that should also not create a new Position

not sure if we want to be that precise though